### PR TITLE
Tests for error callbacks

### PR DESCRIPTION
**Pull request by praveenkumarct**
Added tests for Error callbacks for the Syncers.
- Type
- TaxCategory
- State
- ProductType.

### CODE REVIEW

File: src/integration-test/java/com/commercetools/project/sync/CliRunnerIT.java [hunk 1]
```diff
@@ -110,14 +124,15 @@
...
+  void setup() throws ExecutionException, InterruptedException {
     syncerTestLogger.clearAll();
     cliRunnerTestLogger.clearAll();
     cleanUpProjects(createClient(CTP_SOURCE_CLIENT_CONFIG), createClient(CTP_TARGET_CLIENT_CONFIG));
     setupSourceProjectData(createClient(CTP_SOURCE_CLIENT_CONFIG));
   }
 
-  static void setupSourceProjectData(@Nonnull final SphereClient sourceProjectClient) {
+  static void setupSourceProjectData(@Nonnull final SphereClient sourceProjectClient)
+      throws ExecutionException, InterruptedException {
```
Comment (lojzatran) [inline src/integration-test/java/com/commercetools/project/sync/CliRunnerIT.java:135]: Remove these throws and use `join` instead of `get`.

File: src/main/java/com/commercetools/project/sync/SyncerFactory.java [hunk 2]
```diff
@@ -92,29 +93,31 @@ public static SyncerFactory of(
...
-                          CartDiscountSyncer.of(sourceClient, targetClient, clock)
-                              .sync(runnerNameOptionValue, isFullSync)
-                              .toCompletableFuture(),
-                          InventoryEntrySyncer.of(sourceClient, targetClient, clock)
-                              .sync(runnerNameOptionValue, isFullSync)
-                              .toCompletableFuture(),
-                          CustomerSyncer.of(sourceClient, targetClient, clock)
-                              .sync(runnerNameOptionValue, isFullSync)
-                              .toCompletableFuture());
+              final List<CompletableFuture<Void>> categoryAndInventoryAndCartDiscountAndCustomer =
```
Comment (lojzatran) [inline src/main/java/com/commercetools/project/sync/SyncerFactory.java:96]: Why this change?

File: src/test/java/com/commercetools/project/sync/producttype/ProductTypeSyncerTest.java [hunk 3]
```diff
@@ -67,4 +85,35 @@ void getQuery_ShouldBuildProductTypeQuery() {
     // assertion
     assertThat(query).isEqualTo(ProductTypeReferenceResolutionUtils.buildProductTypeQuery(1));
   }
+
+  @Test
+  void syncWithError_ShouldCallErrorCallback() {
```
Comment (lojzatran) [inline src/test/java/com/commercetools/project/sync/producttype/ProductTypeSyncerTest.java:90]: Please make a better name here including `where` clause. Sth like `syncWithError_WhenNoKeyIsProvided_ShouldCallErrorCallback`. Also update it for other tests.

File: src/test/resources/product-type-without-key.json [hunk 4]
```diff
@@ -0,0 +1,506 @@
+{
+  "id": "cda0dbf7-b42e-40bf-8453-241d5b587f93",
+  "version": 46,
+  "name": "main",
+  "description": "all fleisch",
+  "classifier": "Complex",
+  "attributes": [
```
Comment (lojzatran) [inline src/test/resources/product-type-without-key.json:7]: Just leave one attribute here. This json is too complex for an error case.

File: src/test/resources/shopping-list-no-key.json [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      "productType": {
+        "typeId": "product-type",
+        "id": "eb6ef3da-40a9-4370-8252-33c7b657a016"
+      },
+      "variantId": 2,
+      "quantity": 1,
+      "addedAt": "2020-10-16T07:53:25.917Z"
+    }
+  ],
+  "textLineItems": [
```
Comment (lojzatran) [inline src/test/resources/shopping-list-no-key.json:45]: Remove all `textLineItems`.

File: src/test/resources/shopping-list-no-key.json [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    "en": "shoppingList-name-1"
+  },
+  "description": {
+    "en": "description-1"
+  },
+  "customer": {
+    "typeId": "customer",
+    "id": "368ddf13-cd6c-4ff4-ad38-73038aeed5ba"
+  },
+  "lineItems": [
```
Comment (lojzatran) [inline src/test/resources/shopping-list-no-key.json:15]: Just leave 1 lineItem.

File: src/test/resources/shopping-list-no-key.json [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        },
+        "fields": {
+          "textField": "text line item 1"
+        }
+      },
+      "addedAt": "2020-10-16T09:57:38.551Z"
+    }
+  ],
+  "custom": {
+    "type": {
```
Comment (lojzatran) [inline src/test/resources/shopping-list-no-key.json:84]: Remove this custom field.

File: src/test/resources/tax-category-without-key.json [hunk 8]
```diff
@@ -0,0 +1,41 @@
+{
+  "id": "44e5e889-0878-4c30-941f-767cebebeb7f",
+  "version": 18,
+  "name": "Standard tax category",
+  "rates": [
```
Comment (lojzatran) [inline src/test/resources/tax-category-without-key.json:5]: One rate is enough.

File: src/test/resources/type-without-key.json [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  "name": {
+    "en": "typeName"
+  },
+  "description": {
+    "en": "description"
+  },
+  "resourceTypeIds": [
+    "category"
+  ],
+  "fieldDefinitions": [
```
Comment (lojzatran) [inline src/test/resources/type-without-key.json:11]: One field is enough.

**Review by lojzatran — CHANGES_REQUESTED** [review]
What about tests for error callbacks for `CustomObjectSyncerTest`?

I also see here are lots of code not related to the callbacks. It's hard to review. Next time please only create a PR with the changes you describe.

File: src/integration-test/java/com/commercetools/project/sync/CliRunnerIT.java (hunk 1 above)
Comment (praveenkumarct) [reply src/integration-test/java/com/commercetools/project/sync/CliRunnerIT.java:135]: These changes are not related to this PR.

File: src/main/java/com/commercetools/project/sync/SyncerFactory.java (hunk 2 above)
Comment (praveenkumarct) [reply src/main/java/com/commercetools/project/sync/SyncerFactory.java:96]: These changes are not related to this PR.

File: src/test/java/com/commercetools/project/sync/producttype/ProductTypeSyncerTest.java (hunk 3 above)
Comment (praveenkumarct) [reply src/test/java/com/commercetools/project/sync/producttype/ProductTypeSyncerTest.java:90]: created a new PR and please check this commit for the fix. https://github.com/commercetools/commercetools-project-sync/pull/213/commits/8030e3bdb0a9252deeb2e8a2b519333cc58d0861

File: src/test/resources/product-type-without-key.json (hunk 4 above)
Comment (praveenkumarct) [reply src/test/resources/product-type-without-key.json:7]: created a new PR and please check this commit for the fix. https://github.com/commercetools/commercetools-project-sync/pull/213/commits/8030e3bdb0a9252deeb2e8a2b519333cc58d0861

File: src/test/resources/shopping-list-no-key.json (hunk 5 above)
Comment (praveenkumarct) [reply src/test/resources/shopping-list-no-key.json:45]: Not related to this PR.

File: src/test/resources/shopping-list-no-key.json (hunk 6 above)
Comment (praveenkumarct) [reply src/test/resources/shopping-list-no-key.json:15]: created a new PR and please check this commit for the fix. https://github.com/commercetools/commercetools-project-sync/pull/213/commits/8030e3bdb0a9252deeb2e8a2b519333cc58d0861

File: src/test/resources/shopping-list-no-key.json (hunk 7 above)
Comment (praveenkumarct) [reply src/test/resources/shopping-list-no-key.json:84]: Not related to this PR.

File: src/test/resources/tax-category-without-key.json (hunk 8 above)
Comment (praveenkumarct) [reply src/test/resources/tax-category-without-key.json:5]: created a new PR and please check this commit for the fix. https://github.com/commercetools/commercetools-project-sync/pull/213/commits/8030e3bdb0a9252deeb2e8a2b519333cc58d0861

File: src/test/resources/type-without-key.json (hunk 9 above)
Comment (praveenkumarct) [reply src/test/resources/type-without-key.json:11]: created a new PR and please check this commit for the fix. https://github.com/commercetools/commercetools-project-sync/pull/213/commits/8030e3bdb0a9252deeb2e8a2b519333cc58d0861

File: src/integration-test/java/com/commercetools/project/sync/CliRunnerIT.java (hunk 1 above)
Comment (lojzatran) [reply src/integration-test/java/com/commercetools/project/sync/CliRunnerIT.java:135]: Maybe not anymore in the new PR, but here it was in the PR and if you don't fix them now, they will never be fixed again 😃

Comment (praveenkumarct) [reply src/integration-test/java/com/commercetools/project/sync/CliRunnerIT.java:135]: please also suggest me, where to log the time.
